src/routes/auth/guards/auth.guard.ts

```diff
@@ -0,0 +1,74 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Inject,
+  Injectable,
+} from '@nestjs/common';
+import { IAuthRepository } from '@/domain/auth/auth.repository.interface';
+import { IJwtService } from '@/datasources/jwt/jwt.service.interface';
+import { AuthService } from '@/routes/auth/auth.service';
+import { SiweMessageSchema } from '@/domain/auth/entities/siwe-message.entity';
+
+/**
+ * The AuthGuard should be used to protect routes that require authentication.
+ *
+ * It checks for the presence of a valid JWT access token in the request and
+ * verifies its validity before allowing access to the route.
+ *
+ * 1. Check for the presence of an access token in the request.
+ * 2. Verify the token's validity and decode it.
+ * 3. Ensure the decoded token is that of a valie SiweMessage.
+ * 4. Ensure the token is not expired or not yet valid.
+ * 5. Allow access if all checks pass.
+ */
+
+@Injectable()
+export class AuthGuard implements CanActivate {
+  constructor(
+    @Inject(IAuthRepository) private readonly authRepository: IAuthRepository,
+    @Inject(IJwtService) private readonly jwtService: IJwtService,
+  ) {}
+
+  canActivate(context: ExecutionContext): boolean {
+    const request = context.switchToHttp().getRequest();
+
+    const accessToken = this.authRepository.getAccessToken(
+      request,
+      AuthService.AUTH_TOKEN_TOKEN_TYPE,
+    );
+
+    // No token in the request
+    if (!accessToken) {
+      return false;
+    }
+
+    const payload = ((): unknown | null => {
+      try {
+        return this.jwtService.verify(accessToken);
+      } catch {
+        return null;
+      }
+    })();
+
+    // Token is either not yet valid, expired or malformed
+    if (!payload) {
+      return false;
+    }
+
+    const result = SiweMessageSchema.safeParse(payload);
```

Comment:
Should we validate the schema first? If the schema is not valid from the get go no need to validate the containing properties 🤔

Reply:
Disregard the above as the `payload` is only retrieved via the `jwtService` so this is correct 👍